File: packages/typescript/src/modules/reconciliation.ts
/**
 * ReconciliationModule: transaction status query and polling reconciliation.
 */

import { mapReturnCodeToError } from '../errors/index.js';
import type { EcollectConfig } from '../config.js';
import { TRANSACTION_INFO_URLS } from '../config.js';
import type {
  GetTransactionInformationRequest,
  TransactionInfoResponse,
  PaymentInfoItem,
} from '../types/api.js';
import type { TransactionResult } from '../types/index.js';
import type { HttpClient } from '../utils/http.js';
import type { SessionModule } from './session.js';
import { PollingManager } from '../utils/polling.js';

function mapTxInfo(res: TransactionInfoResponse): TransactionResult {
  return {
    returnCode: res.ReturnCode,
    ticketId: res.TicketId,
    tranState: res.TranState as TransactionResult['tranState'],
    trazabilityCode: res.TrazabilityCode,
    transValue: res.TransValue,
    transVatValue: res.TransVatValue,
    payCurrency: res.PayCurrency,
    currencyRate: res.CurrencyRate,
    bankProcessDate: res.BankProcessDate,
    fiCode: res.FICode,
    fiName: res.FiName,
    paymentSystem: res.PaymentSystem,
    tranCycle: res.TransCycle,
    invoice: res.Invoice,
    referenceArray: res.ReferenceArray,
    srvCode: res.SrvCode,
    paymentInfoArray: res.PaymentInfoArray?.map((i) => ({
      code: i.AttributeCode,
      desc: i.AttributeDesc,
      value: i.AttributeValue,
    })),
  };
}

export class ReconciliationModule {
  private readonly config: EcollectConfig;
  private readonly http: HttpClient;
  private readonly session: SessionModule;
  private readonly pollingManager: PollingManager;

  constructor(config: EcollectConfig, http: HttpClient, session: SessionModule) {
    this.config = config;
    this.http = http;
    this.session = session;
    this.pollingManager = new PollingManager((ticketId) => this.getTransactionStatus(ticketId));
  }

  /**
   * Query the current state of a transaction.
   * Optionally pass merchantTransactionId as fallback.
   */
  async getTransactionStatus(
    ticketId: number,
    merchantTransactionId?: string,
  ): Promise<TransactionResult> {
    const url = TRANSACTION_INFO_URLS[this.config.environment];
    const sessionToken = await this.session.getActive();

    const paymentInfoArray: PaymentInfoItem[] | undefined = merchantTransactionId
      ? [{ AttributeCode: 26, AttributeDesc: 'MerchantTransactionId', AttributeValue: merchantTransactionId }]
      : undefined;

    const body: GetTransactionInformationRequest = {
      EntityCode: this.config.etyCode,
      SessionToken: sessionToken,
      TicketId: ticketId,
      PaymentInfoArray: paymentInfoArray,
    };

    const res = await this.http.post<GetTransactionInformationRequest, TransactionInfoResponse>(
      url,
      body,
    );

    if (res.ReturnCode === 'FAIL_APIEXPIREDSESSION') {
      this.session.invalidate();
      body.SessionToken = await this.session.getActive();
      const retried = await this.http.post<GetTransactionInformationRequest, TransactionInfoResponse>(
        url,
        body,
      );
      if (retried.ReturnCode !== 'SUCCESS') {
        throw mapReturnCodeToError(retried.ReturnCode);
      }
      return mapTxInfo(retried);
    }

    if (res.ReturnCode !== 'SUCCESS') {
      throw mapReturnCodeToError(res.ReturnCode);
    }

    return mapTxInfo(res);
  }

  /**
   * Start background polling for a transaction until a final state is reached.
   * @param ticketId - The ecollect ticket ID
   * @param timeout - Timeout in ms (default: 10 minutes)
   * @returns Promise that resolves to the final TransactionResult
   */
  async reconciliate(ticketId: number, timeout = 600_000): Promise<TransactionResult> {
    return this.pollingManager.waitForFinalState(ticketId, timeout);
  }

  /**
   * Stop an in-progress reconciliation poll.
   */
  stopReconciliation(ticketId: number): void {
    this.pollingManager.stopPolling(ticketId);
  }
}

File: packages/python/ecollect/modules/reconciliation.py
"""Reconciliation module: transaction status queries and polling-based resolution."""
import logging
from typing import Any, Callable, Dict, Optional

from ecollect.config import EcollectConfig
from ecollect.exceptions import raise_for_return_code
from ecollect.utils.http import AsyncHttpClient, run_sync
from ecollect.utils.polling import PollingManager

logger = logging.getLogger(__name__)


class ReconciliationModule:
    """Query transaction status and reconcile pending transactions."""

    def __init__(
        self,
        config: EcollectConfig,
        http: AsyncHttpClient,
        session: "SessionModule",  # type: ignore[name-defined]
    ) -> None:
        self._config = config
        self._http = http
        self._session = session
        self._polling = PollingManager()

    async def _post(self, endpoint: str, payload: Dict[str, Any]) -> Dict[str, Any]:
        url = self._config.endpoint_url(endpoint)
        data = await self._http.post_with_retry(url, payload)
        return_code = data.get("ReturnCode", "")
        if return_code == "FAIL_APIEXPIREDSESSION":
            self._session.invalidate()
            payload["SessionToken"] = await self._session.get_active()
            data = await self._http.post_with_retry(url, payload)
        raise_for_return_code(data.get("ReturnCode", ""), str(data))
        return data

    # ------------------------------------------------------------------
    # Async API
    # ------------------------------------------------------------------

    async def get_transaction_status(
        self,
        ticket_id: int,
        merchant_transaction_id: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Query the current state of a transaction.

        Args:
            ticket_id: ecollect TicketId.
            merchant_transaction_id: Optional fallback if TicketId is unavailable.
        """
        session_token = await self._session.get_active()
        payload: Dict[str, Any] = {
            "EntityCode": self._config.ety_code,
            "SessionToken": session_token,
            "TicketId": ticket_id,
        }
        if merchant_transaction_id:
            payload["PaymentInfoArray"] = [
                {
                    "AttributeCode": 26,
                    "AttributeDesc": "MerchantTransactionId",
                    "AttributeValue": merchant_transaction_id,
                }
            ]
        return await self._post("getTransactionInformation", payload)

    async def reconciliate(
        self,
        ticket_id: int,
        on_complete: Optional[Callable[[Dict[str, Any]], None]] = None,
        timeout: int = 600,
    ) -> Dict[str, Any]:
        """Poll transaction status until a terminal state is reached.

        Args:
            ticket_id: ecollect TicketId to reconcile.
            on_complete: Optional callback called with the final status.
            timeout: Seconds before raising PollingTimeoutError.

        Returns:
            The final transaction status dict.
        """
        result: Dict[str, Any] = {}

        def _on_complete(status: Dict[str, Any]) -> None:
            nonlocal result
            result = status
            if on_complete:
                on_complete(status)

        await self._polling.start_polling(
            ticket_id=ticket_id,
            get_status=self.get_transaction_status,
            on_complete=_on_complete,
            timeout=timeout,
        )
        return result

    # ------------------------------------------------------------------
    # Sync wrappers
    # ------------------------------------------------------------------

    def get_transaction_status_sync(
        self,
        ticket_id: int,
        merchant_transaction_id: Optional[str] = None,
    ) -> Dict[str, Any]:
        return run_sync(self.get_transaction_status(ticket_id, merchant_transaction_id))

    def reconciliate_sync(
        self,
        ticket_id: int,
        on_complete: Optional[Callable[[Dict[str, Any]], None]] = None,
        timeout: int = 600,
    ) -> Dict[str, Any]:
        return run_sync(self.reconciliate(ticket_id, on_complete, timeout))
